packages/data-stores/src/site/selectors.ts

```diff
@@ -2,10 +2,28 @@
  * Internal dependencies
  */
 import { State } from './reducer';
+import { dispatch } from '@wordpress/data';
+import { STORE_KEY } from './constants';
 
 export const getState = ( state: State ) => state;
 
 export const getNewSite = ( state: State ) => state.newSite.data;
 export const getNewSiteError = ( state: State ) => state.newSite.error;
 export const isFetchingSite = ( state: State ) => state.newSite.isFetching;
 export const isNewSite = ( state: State ) => !! state.newSite.data;
+
+/**
+ * Get a site matched by id. This selector has a matching
+ * resolver that uses the `siteId` parameter to fetch an existing site. If the
+ * site cannot be found, invalidate the resolution cache.
+ *
+ * @param state {State}		state object
+ * @param siteId {number}	id of the site to look up
+ */
+export const getSite = ( state: State, siteId: number ) => {
+	const site = state.sites[ siteId ];
+	if ( ! site ) {
+		dispatch( 'core/data' ).invalidateResolution( STORE_KEY, 'getSite', [ siteId ] );
```

Comment:
On second look -- can we keep the selector side-effect free, and move the invalidation into the resolver?

Reply:
I think that'd be the ideal way to handle it. I had a go at doing it that way before moving it to the selector but ran into the following problem:

In the metadata `fulfillSelector` function that's called to manage resolving a selector, it [dispatches a `finishResolution` action](https://github.com/wordpress/gutenberg/blob/4857ad58c1241b3d63d21a6880c989b85746c3dc/packages/data/src/namespace-store/index.js#L267) after running `fulfillResolver`. This means that if we attempt to dispatch actions within the resolver to invalidate the resolution, the value will be overridden as soon as the resolver finishes resolving.

This is unfortunate because it means we can't use the cool resolvers cache middleware which appears to be designed for use cases quite similar to ours, where we'd add a [`shouldInvalidate` method to our reducer](https://github.com/wordpress/gutenberg/blob/4857ad58c1241b3d63d21a6880c989b85746c3dc/packages/data/src/resolvers-cache-middleware.js#L40) to be evaluated based on a particular action.

The way I'd _like_ to implement this would be something like:

```ts
// resolvers.ts

export function* getSite( siteId: number ) {
	try {
		const existingSite = yield wpcomRequest( {
			path: '/sites/' + encodeURIComponent( siteId ),
			apiVersion: '1.1',
		} );
		yield dispatch( STORE_KEY ).receiveSite( siteId, existingSite );
	} catch ( err ) {
		// Finish resolution early to enable firing off actions after resolution is complete
		yield dispatch( 'core/data' ).finishResolution( STORE_KEY, 'getSite', [ siteId ] ); 
		// The following action will be matched in `shouldInvalidate` so that the middleware will invalidate the cache
		yield dispatch( STORE_KEY ).receiveSiteFailed( siteId, undefined );
	}
}

getSite.shouldInvalidate = ( action, siteId ) => {
	if ( action.type === 'RECEIVE_SITE_FAILED' && action.siteId === siteId) {
		return true;
	}
	return false;
}
```

This nearly works, the blocker is that the after running the above, the `fulfillSelector` function fires off that `finishResolution` action again, setting the resolver as finished and cached (`false`).

I think an upstream fix to enable the above pattern (which is otherwise used by actions dispatched outside of resolvers) would be to change the `subKeysIsResolved` reducer so that [the `FINISH_RESOLUTION` action](https://github.com/wordpress/gutenberg/blob/4857ad58c1241b3d63d21a6880c989b85746c3dc/packages/data/src/namespace-store/metadata/reducer.js#L27) only sets the state to `false` if the state was previously `true` (it is currently resolving). As an aside to this PR, does that sound like something that is generalisable for an upstream PR to you? (I'm happy to pursue it further because I'd love to enable a pattern where we can invalidate within the reducer and / or queue actions to be dispatched after resolution)

All that said, where it leaves us in this PR is that I think our only practical option is to call `invalidateResolution` from the selector, unless you can think of another approach?

Reply:
Thanks for this detailed explanation! (And sorry for taking a while to get back to this :sweat_smile: )

> The way I'd _like_ to implement this would be something like: [...]

This sounds good (especially using `shouldInvalidate`, tho I'm thinking that in an ideal world, we shouldn't even have to call `yield dispatch( 'core/data' ).finishResolution( STORE_KEY, 'getSite', [ siteId ] ); ` manually, should we? Ideally, this should be the responsibility of the underlying mechanism.

> I think an upstream fix to enable the above pattern (which is otherwise used by actions dispatched outside of resolvers) would be to change the `subKeysIsResolved` reducer so that [the `FINISH_RESOLUTION` action](https://github.com/wordpress/gutenberg/blob/4857ad58c1241b3d63d21a6880c989b85746c3dc/packages/data/src/namespace-store/metadata/reducer.js#L27) only sets the state to `false` if the state was previously `true` (it is currently resolving). As an aside to this PR, does that sound like something that is generalisable for an upstream PR to you? (I'm happy to pursue it further because I'd love to enable a pattern where we can invalidate within the reducer and / or queue actions to be dispatched after resolution)

I think that would make sense upstream! It's not really tied to our problem domain but seems more general -- it's conceivable that users of a resolver would like to invalidate the resolution cache upon network error regardless of what the corresponding state actually holds.

And it actually sounds like a bug that there's a `shouldInvalidate` method that can be added to resolvers, but that gets overridden by `finishResolution`.

> All that said, where it leaves us in this PR is that I think our only practical option is to call `invalidateResolution` from the selector, unless you can think of another approach?

I don't see any option at this point (and I think you have an overall better grasp of resolver cache invalidation than I do :sweat_smile: ).

Reply:
Thanks for following up @ockham and not a worry at all, we've been pretty busy this past week! I'll start having a play in Gutenberg and see how I might propose something upstream. Will let you know when I've put together something to look at :)